pm: Update API and subsystem to support SMP

ceolin opened the pull request:
- ~Change the power management API that is implemented by the
  architecture adding a parameter with which CPU is idle.~
- Update policy API to handle multiple CPUs
  - Change the residency policy to get power states per CPUs and properly use them
- Track power states per core (instead of a global state)

Changes from initial version:

- Properly handle device power management on SMP environment
- Use unsigned int to reference a cpu
- Remove `_handle_device_abort` . It was only called in one place and it was simple enough to be inline in the caller.

Commit:
pm: Change post_ops_done type to boolean

Boolean is more appropriate to this variable and it is aligned with
Zephyr's code guideline.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Track power state per cpu

Currently power management subsystem uses a global variable to hold
power state. On multicore environment this is a problem and we have to
have this information per core.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: policy: Add cpu information in the API

On multicore environments the policy may need to know which CPU is
idle.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: policy: residency: Support multiple CPUs

Each CPU can have its own set of states. Update the policy to get them
from DT and account it when return the best state to the kernel.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

teburd reviewed (changes requested):
Looks good, I want to understand what the change in sched.c was for before approving though

teburd reviewed (approved):
LGTM

Commit:
pm: Document missing parameter in exit post ops

info parameter was not properly documented.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Change pm_power_state_next_get to support SMP

Change the API to account multiple CPUs.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Better return type for pm_system_suspend

Instead of returning PM_STATE_ACTIVE for when the cpu didn't enter a
low power state and a different state when it entered, but has
already left the state and is active again, it changes
pm_system_suspend to return true when the cpu has entered a low power
state and false otherwise.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

peter-mitsis commented on include/pm/pm.h:
```diff
@@ -72,10 +72,11 @@ struct pm_notifier {
  * @param info Power state which should be used in the ongoing
  *	suspend operation.
  */
-void pm_power_state_force(struct pm_state_info info);
+void pm_power_state_force(int cpu, struct pm_state_info info);
```
I observe that the [cpu] parameter is using the signed integer type. Would it be both better and more consistent to use an unsigned integer type? I ask this for the following reasons ...

1. A negative CPU ID does not make much immediate sense to me. (I may of course be missing a use case.)
2. The CPU id stored within "struct _cpu" is of type uint8_t. So keeping the unsigned quality helps avoid signed/unsigned mix-and-match.

ceolin replied:
I followed some API that use `int` for this purpose, but I really think you are right about it.

peter-mitsis commented on subsys/pm/power.c:
```diff
@@ -173,8 +173,10 @@ void pm_system_resume(void)
 	 * notification is not required.
 	 */
 	if (!post_ops_done) {
+		int cpu = _current_cpu->id;
+
```
Since _current_cpu->id is an uint8_t, it may be preferrable to keep the unsigned property with [cpu] as well.

peter-mitsis commented on include/pm/policy.h:
```diff
@@ -20,11 +20,12 @@ extern "C" {
- * @return The power state the system should use.
+ * @return The power state the system should use for the given cpu.
  */
-struct pm_state_info pm_policy_next_state(int32_t ticks);
+struct pm_state_info pm_policy_next_state(int cpu, int32_t ticks);
```
At the risk of sounding like a broken record, again, I would suggest using an unsigned type over a signed type. I will also refrain from repeating this same comment in subsequents commits in this PR--if applicable.

It is also not a hill that I am willing to die on. :)

Commit:
pm: smp: Properly handle device pm on SMP

On smp environment, we cannot do device power management based only on
the state the current core. For example, do not suspend devices if
there are other cores active, or resume devices if there are other
cores already active.

This change ensure that devices are only suspend when the last active
core become idle and that these devices are resumed only when the first
core becomes active.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Remove _handle_device_abort

_handle_device_abort is only called in one place and it basically does
two instructions. Just remove it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Keep cpu power state consistent

Set the cpu state to ACTIVE after it resumes.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Keep track of active states on state_force

Update z_cpu_active in pm_power_state_force

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@teburd @peter-mitsis  @gmarull  can you take a look on it again ?

ceolin commented on subsys/pm/power.c:
```diff
@@ -249,7 +251,14 @@ void pm_power_state_force(struct pm_state_info info)
-	z_power_states[_current_cpu->id] = info;
+	z_power_states[id] = info;
+
+#ifdef PM_DEVICE_STATE_ACTIVE
```
wrong word complete, will fix it

ceolin replied:
It should be `CONFIG_PM_DEVICE` since it is only required when it is enabled

ceolin commented on subsys/pm/policy/policy_residency.c:
```diff
@@ -4,48 +4,72 @@
+	DT_FOREACH_CHILD(DT_PATH(cpus), CPU_STATES_SIZE)
+};
+#else
+static const struct pm_state_info *pm_min_residency[CONFIG_MP_NUM_CPUS] = {};
+static int pm_min_residency_sizes[CONFIG_MP_NUM_CPUS] = {};
```
BUILD_ASSERT added to enforce cpus node.

ceolin replied:
I reverted the change. If I add the BUILD_ASSERT, all native_posix because it does not implement it. For now I think it is fine to keep the old behavior. I don't want to just put `HAS_NO_PM` for native posix because many pm tests are using native posix only.

teburd commented on subsys/pm/power.c:
```diff
@@ -226,14 +229,20 @@ void pm_system_resume(void)
 	}
 }
 
 void pm_power_state_force(struct pm_state_info info)
```
Maybe I'm wrong, but I feel as if this should take the cpu id as a parameter to force the state on here, otherwise the API and docs make it seems like the call would be a system wide power state and not for the currently cpu context

There may be cases as well where a task on one cpu decides when to force the other to a particular state.

I get this has some maybe unwanted implications, but it does feel like the better API to have. Maybe that needs to be built on top of this though, so that you schedule a task to run this function on the cpu you want to force a power state on.

ceolin replied:
I will remove this function. Will do it in a different PR to bring more visibility for this. The force behind this is that this API is very problematic in SMP environments. For example, one cpu may suspend another one that was in holding a mutex or spin lock that will cause a problem to the system. In this situation what should be done ? abort the thread running in that core ? just scheduled out the thread ? but if it is pinned to the suspended core ?

teburd commented on subsys/pm/power.c:
```diff
@@ -320,21 +334,21 @@ enum pm_state pm_system_suspend(int32_t ticks)
 #if CONFIG_PM_DEVICE
-	if (should_resume_devices) {
-		/* Turn on peripherals and restore device states as necessary */
+	if (atomic_add(&z_cpu_active, 1) == 0) {
```
:+1:

teburd commented on subsys/pm/power.c:
```diff
@@ -242,69 +251,74 @@ void pm_power_state_force(struct pm_state_info info)
-
-	if (z_power_state.state != PM_STATE_RUNTIME_IDLE) {
+	if ((z_power_states[id].state != PM_STATE_RUNTIME_IDLE) &&
+			(atomic_sub(&z_cpu_active, 1) == 1)) {
```
:+1:

teburd commented on subsys/pm/power.c:
```diff
@@ -20,9 +20,12 @@
 static sys_slist_t pm_notifiers = SYS_SLIST_STATIC_INIT(&pm_notifiers);
+static struct pm_state_info z_power_states[CONFIG_MP_NUM_CPUS];
+#ifdef CONFIG_PM_DEVICE
+static atomic_t z_cpu_active = ATOMIC_INIT(CONFIG_MP_NUM_CPUS);
```
nit: z_cpu_active might read easier if pluralized (z_cpus_active)

teburd commented on tests/subsys/pm/device_runtime_api/prj.conf:
```diff
@@ -1,4 +1,3 @@
 CONFIG_ZTEST=y
-CONFIG_PM=y
```
why the config change here?

teburd reviewed (approved):
I think this looks good, and the atomic cpu active count makes good sense, some nits and comments that might be addressed in future improvements
